### PR TITLE
feat(sdk): Support return values in lightweight Python function components v2.

### PR DESCRIPTION
**Pull request by neuromage**
Adds support for return values (scalars, artifacts and named tuples of these) to lightweight function components in v2.

**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

### CODE REVIEW

**Comment by Ark-kun** [comment]
Thank you for this improvement. 
Reviewing it now.
/assign @Ark-kun

File: sdk/python/kfp/components/executor.py [hunk 1]
```diff
@@ -182,7 +183,7 @@ def _write_executor_output(self, func_output: Optional[Any] = None):
                   len(self._return_annotation._fields), self._func.__name__,
                   len(func_output)))
         for field in self._return_annotation._fields:
-          field_type = self._return_annotation._field_types[field]
+          field_type = self._return_annotation.__annotations__[field]
```
Comment (Ark-kun) [inline sdk/python/kfp/components/executor.py:186]: Is this compatible with both Python v3.6 and v3.9?

File: sdk/python/kfp/components/_python_op.py [hunk 2]
```diff
@@ -535,6 +538,15 @@ def _func_to_component_spec_v2(
 
         arguments.extend(arguments_for_input)
 
+    # Add output placeholders for return values from func.
+    func_outputs = [
+        output for output in component_outputs
+        if output._passing_style is None
+    ]
+    for output in func_outputs:
```
Comment (Ark-kun) [inline sdk/python/kfp/components/_python_op.py:546]: I think this should already been handled here:

https://github.com/kubeflow/pipelines/blob/46cd8196b0297b0af9966bd20dad4ddeac7142a5/sdk/python/kfp/components/_python_op.py#L535

If it's not, then there is some issue with io_types.OutputArtifact handling somewhere else.

Comment (Ark-kun) [inline sdk/python/kfp/components/_python_op.py:546]: JFYI: In the current lightweight components implementation, the paths for the by-value outputs are passed through the `----output-paths` argument which lists the paths.

File: sdk/python/kfp/components/_python_op.py [hunk 3]
```diff
@@ -535,6 +538,15 @@ def _func_to_component_spec_v2(
...
+        flag = "--" + output.name.replace("_", "-")
```
Comment (Ark-kun) [inline sdk/python/kfp/components/_python_op.py:547]: 1) How do we prevent name clashes? For example, there can be input called `model` and output called `model`?
2) Do we want to add the `-path` suffix to the name to make the names more readable?

```
--loss /path/to/loss/file
```
vs.
```
--loss-path /path/to/loss/file
```

Adding `-output-path` suffix can solve both of those issues:
```
--loss-output-path /path/to/loss/file
```

WDYT?

File: sdk/python/kfp/components/executor.py [hunk 4]
```diff
@@ -41,13 +41,16 @@ def __init__(self, executor_input: Dict, function_to_execute: Callable):
         self._output_artifacts[name] = self._make_output_artifact(
             artifacts_list[0])
 
+    self._return_annotation = inspect.signature(self._func).return_annotation
+    self._executor_output = {}
+
   @classmethod
-  def _make_input_artifact(self, runtime_artifact: Dict):
+  def _make_input_artifact(cls, runtime_artifact: Dict):
```
Comment (Ark-kun) [inline sdk/python/kfp/components/executor.py:48]: Should this be a static method or maybe just a normal function?

File: sdk/python/kfp/components/executor.py [hunk 5]
```diff
@@ -98,8 +101,64 @@ def _get_input_artifact_path(self, artifact_name: str):
...
+      f.write(str(value))
+
+  @classmethod
+  def _is_parameter(cls, annotation: Any) -> bool:
+    return annotation in [str, int, float]
+
+  @classmethod
+  def _is_artifact(cls, annotation: Any) -> bool:
+    if type(annotation) == type:
+      return issubclass(annotation, Artifact)
```
Comment (Ark-kun) [inline sdk/python/kfp/components/executor.py:134]: How does this work for custom artifact types?

File: sdk/python/kfp/components/executor.py [hunk 6]
```diff
@@ -98,8 +101,64 @@ def _get_input_artifact_path(self, artifact_name: str):
...
     return input_artifact.path
 
-  def _write_executor_output(self):
-    executor_output = {'artifacts': {}}
+  def _write_output_parameter_value(self, name: str, value: Union[str, int,
+                                                                  float]):
+    if type(value) == str:
+      output = {"stringValue": value}
+    elif type(value) == int:
+      output = {"intValue": value}
```
Comment (Ark-kun) [inline sdk/python/kfp/components/executor.py:109]: Are integers planned to be supported in MLMD?

File: sdk/python/kfp/components/executor_types.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+import json
+import inspect
+from typing import Callable, Dict
+from kfp.components._python_op import RuntimeArtifact, InputArtifact, OutputArtifact, InputPath, OutputPath
+
+
+class Executor():
```
Comment (Ark-kun) [inline sdk/python/kfp/components/executor_types.py:7]: You've previously added class `Executor` in `executor.py`. Can we explain the main motivation for having two classes with same names to the user?

File: sdk/python/kfp/components/_python_op.py (hunk 3 above)
Comment (neuromage) [reply sdk/python/kfp/components/_python_op.py:547]: Done.

File: sdk/python/kfp/components/executor.py (hunk 4 above)
Comment (neuromage) [reply sdk/python/kfp/components/executor.py:48]: No. Classmethods let me easily include it in lightweight components by source. Static methods are discouraged: https://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators

File: sdk/python/kfp/components/executor.py (hunk 6 above)
Comment (neuromage) [reply sdk/python/kfp/components/executor.py:109]: This is how it's written in ExecutorOutput.

File: sdk/python/kfp/components/executor.py (hunk 5 above)
Comment (neuromage) [reply sdk/python/kfp/components/executor.py:134]: It's not supported right now.

File: sdk/python/kfp/components/executor.py (hunk 1 above)
Comment (neuromage) [reply sdk/python/kfp/components/executor.py:186]: Yes, I believe it is.

File: sdk/python/kfp/components/executor_types.py (hunk 7 above)
Comment (neuromage) [reply sdk/python/kfp/components/executor_types.py:7]: Old file, accidentally committed. Removed.

File: sdk/python/kfp/components/executor.py (hunk 6 above)
Comment (Ark-kun) [reply sdk/python/kfp/components/executor.py:109]: JFYI: I've heard there are some plans to change ExecutorOutput.

File: sdk/python/kfp/components/executor_types.py (hunk 7 above)
Comment (Ark-kun) [reply sdk/python/kfp/components/executor_types.py:7]: //You might want to push the latest changes.

Comment (neuromage) [reply sdk/python/kfp/components/executor_types.py]: Done.

File: sdk/python/kfp/v2/compiler_cli_tests/compiler_cli_tests.py [hunk 8]
```diff
@@ -23,24 +23,33 @@
 
 class CompilerCliTests(unittest.TestCase):
 
-  def _test_compile_py_to_json(self, file_base_name, additional_arguments=[]):
+  def _test_compile_py_to_json(self,
+                               file_base_name,
+                               additional_arguments=[],
+                               update_golden: bool = False):
```
Comment (chensun) [inline sdk/python/kfp/v2/compiler_cli_tests/compiler_cli_tests.py:29]: JFYI, I usually just execute the test pipeline `.py` file to update golden. This works for all recent samples:

https://github.com/kubeflow/pipelines/blob/4c51b576f577e337049599cb85b098006664265f/sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_condition.py#L46-L50

Earlier samples have `output_path=__file__ + '.json'` instead, so I could take a look a the updated golden before renaming it to replace the existing one. But later I found that extra step is not really necessary.

**Review by chensun — APPROVED** [review]
/lgtm 

Thanks!

File: sdk/python/kfp/v2/compiler_cli_tests/compiler_cli_tests.py (hunk 8 above)
Comment (neuromage) [reply sdk/python/kfp/v2/compiler_cli_tests/compiler_cli_tests.py:29]: I see, got it, thanks!

**Pull request merged**